### PR TITLE
Adds generic factory overloads with both TService and TImplementation for TryAdd*

### PR DESCRIPTION
**Pull request by khellang**
This sits on top of #353

### CODE REVIEW

File: src/Microsoft.Extensions.DependencyInjection.Abstractions/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -504,10 +544,7 @@ public static void TryAddSingleton<TService>(this IServiceCollection collection,
                 throw new ArgumentNullException(nameof(descriptors));
             }
 
-            foreach (var d in descriptors)
-            {
-                services.TryAddEnumerable(d);
-            }
+            return descriptors.All(services.TryAdd);
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/Extensions/ServiceCollectionExtensions.cs]: I'm not sure this is equivalent.

Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/Extensions/ServiceCollectionExtensions.cs]: Oh, good catch!

Comment (henkmollema) [inline src/Microsoft.Extensions.DependencyInjection.Abstractions/Extensions/ServiceCollectionExtensions.cs]: That's what I meant in the other PR

**Comment by khellang** [comment]
I guess someone else can PR this if they want it later.

**Pull request closed without merging**